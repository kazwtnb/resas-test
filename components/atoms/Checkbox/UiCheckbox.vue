<template>
  <div class="ui-checkbox" @click="onClick($event)">
    <label>
      <input type="checkbox" :checked="modelValue" />
      <div
        class="ui-checkbox-box"
        :class="{ 'ui-checkbox-box--checked': modelValue }"
      >
        <i class="fa-solid fa-check" />
      </div>
      <span>{{ label }}</span>
    </label>
  </div>
</template>
<script setup lang="ts">
interface Props {
  modelValue?: boolean;
  label: string;
}

interface Emits {
  (e: 'update:modelValue', value: boolean): void;
  (e: 'click', value: MouseEvent): void;
}

const props = withDefaults(defineProps<Props>(), {
  modelValue: false
});
const emit = defineEmits<Emits>();

const onClick = (event: MouseEvent): void => {
  event.preventDefault();
  emit('update:modelValue', !props.modelValue);
  emit('click', event);
};
</script>
<style lang="scss" scoped>
.ui-checkbox {
  display: inline-block;

  label {
    display: inline-flex;
    align-items: center;
    cursor: pointer;
  }

  input {
    border: 0;
    clip: rect(0 0 0 0);
    height: 1px;
    width: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    white-space: nowrap;
    pointer-events: none;
  }

  .ui-checkbox-box {
    display: inline-flex;
    width: 1rem;
    height: 1rem;
    align-items: center;
    justify-content: center;
    border: 1px solid colors.$color-border;
    border-radius: 2px;
    color: colors.$color-border;
    font-size: 0.75rem;
    transition: color 200ms ease-in, background-color 200ms ease-in;

    &:not(.ui-checkbox-box--checked) {
      > i {
        opacity: 0.3;
      }
    }

    &.ui-checkbox-box--checked {
      color: colors.$color-white;
      background-color: colors.$color-primary;
      border-color: colors.$color-primary;
    }

    + * {
      margin-left: 0.5rem;
    }
  }

  &:hover .ui-checkbox-box,
  input:focus-visible + .ui-checkbox-box {
    &:not(.ui-checkbox-box--checked) {
      background-color: colors.$color-hover;
    }

    &.ui-checkbox-box--checked {
      background-color: colors.$color-primary-hover;
      border-color: colors.$color-primary-hover;
    }
  }
}
</style>
