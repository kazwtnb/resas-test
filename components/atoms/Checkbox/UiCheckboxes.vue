<template>
  <ul class="ui-checkboxes">
    <li v-for="item of props.items" :key="idKey ? item[idKey] : item">
      <UiCheckbox
        :model-value="selected(item)"
        :label="labelKey ? item[labelKey] : item"
        @click="onClick($event, item)"
      />
    </li>
  </ul>
</template>
<script setup lang="ts">
// Nuxt3のAutoImportがテストで効かないため明示的にimportを記載
import { toRefs } from 'vue';
import { useMultiSelect } from '~/composables/useMultiselect';
import UiCheckbox from '~/components/atoms/Checkbox/UiCheckbox.vue';

interface Props<T> {
  modelValue?: T[];
  items: T[];
  idKey?: string;
  labelKey?: string;
  minItemWidth?: string;
}

interface Emits<T> {
  (e: 'update:modelValue', value: T[]): void;
  (e: 'change'): void;
  (e: 'add', value: T): void;
  (e: 'remove', value: T): void;
}

// VueでGenericsのprops、emitsの定義が不可なためanyを使用して回避
/* eslint-disable @typescript-eslint/no-explicit-any */

const props = withDefaults(defineProps<Props<any>>(), {
  modelValue: () => [],
  idKey: '',
  labelKey: '',
  minItemWidth: '90px'
});
const { modelValue } = toRefs(props);
const { selected, toggle } = useMultiSelect(modelValue, props.idKey);
const emit = defineEmits<Emits<any>>();

const onClick = (event: MouseEvent, item: any): void => {
  event.preventDefault();
  const result = toggle(item);

  emit('update:modelValue', result.value);
  emit('change');

  switch (result.type) {
    case 'add':
      emit('add', item);
      break;
    case 'remove':
      emit('remove', item);
      break;
  }
};
</script>
<style lang="scss" scoped>
ul.ui-checkboxes {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(v-bind(minItemWidth), 1fr));
  grid-gap: 1rem;
}
</style>
