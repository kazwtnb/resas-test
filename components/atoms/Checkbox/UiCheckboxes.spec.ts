import { mount } from '@vue/test-utils';
import { describe, expect, test } from 'vitest';
import UiCheckboxes from './UiCheckboxes.vue';
import UiCheckbox from './UiCheckbox.vue';

describe('UiCheckboxes', () => {
  const object1 = { id: '1', value: 'a' };
  const object2 = { id: '2', value: 'b' };

  describe('チェックボックスクリック時に正しいイベントがemitされること', () => {
    test.each([
      {
        name: 'チェックを入れる',
        input: {
          modelValue: [],
          items: [object1],
          clickIndex: 0
        },
        expected: [
          { event: 'update:modelValue', value: [object1] },
          { event: 'change' },
          { event: 'add', value: object1 }
        ]
      },
      {
        name: 'チェックをはずす',
        input: {
          modelValue: [object1],
          items: [object1],
          clickIndex: 0
        },
        expected: [
          { event: 'update:modelValue', value: [] },
          { event: 'change' },
          { event: 'remove', value: object1 }
        ]
      }
    ])(`%o`, ({ input, expected }) => {
      const wrapper = mount(UiCheckboxes, {
        props: {
          modelValue: input.modelValue,
          items: input.items,
          idKey: 'id',
          labelKey: 'value'
        }
      });

      wrapper.findAllComponents(UiCheckbox)[input.clickIndex].trigger('click');

      expected.forEach((e) => {
        if (e.value) {
          expect(wrapper.emitted(e.event)).toEqual([[e.value]]);
          return;
        }

        expect(wrapper.emitted(e.event)).toBeTruthy();
      });
    });
  });

  describe('snapshot', () => {
    test.each([
      {
        name: '未選択',
        modelValue: [],
        items: [object1, object2]
      },
      {
        name: '1つ目を選択済み',
        modelValue: [object1],
        items: [object1, object2]
      },
      {
        name: 'すべて選択済み',
        modelValue: [object1, object2],
        items: [object1, object2]
      }
    ])(`%o`, (input) => {
      const wrapper = mount(UiCheckboxes, {
        props: {
          modelValue: input.modelValue,
          items: input.items,
          idKey: 'id',
          labelKey: 'value'
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
