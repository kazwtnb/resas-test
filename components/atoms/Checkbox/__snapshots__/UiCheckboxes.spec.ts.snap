// Vitest Snapshot v1

exports[`UiCheckboxes > snapshot > {
  name: '1つ目を選択済み',
  modelValue: [ { id: '1', value: 'a' }, [length]: 1 ],
  items: [ { id: '1', value: 'a' }, { id: '2', value: 'b' }, [length]: 2 ]
} 1`] = `
<ul
  class="ui-checkboxes"
  data-v-4ad4a939=""
>
  
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box ui-checkbox-box--checked"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          a
        </span>
      </label>
    </div>
  </li>
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          b
        </span>
      </label>
    </div>
  </li>
  
</ul>
`;

exports[`UiCheckboxes > snapshot > {
  name: 'すべて選択済み',
  modelValue: [ { id: '1', value: 'a' }, { id: '2', value: 'b' }, [length]: 2 ],
  items: [ { id: '1', value: 'a' }, { id: '2', value: 'b' }, [length]: 2 ]
} 1`] = `
<ul
  class="ui-checkboxes"
  data-v-4ad4a939=""
>
  
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box ui-checkbox-box--checked"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          a
        </span>
      </label>
    </div>
  </li>
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box ui-checkbox-box--checked"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          b
        </span>
      </label>
    </div>
  </li>
  
</ul>
`;

exports[`UiCheckboxes > snapshot > {
  name: '未選択',
  modelValue: [ [length]: 0 ],
  items: [ { id: '1', value: 'a' }, { id: '2', value: 'b' }, [length]: 2 ]
} 1`] = `
<ul
  class="ui-checkboxes"
  data-v-4ad4a939=""
>
  
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          a
        </span>
      </label>
    </div>
  </li>
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          b
        </span>
      </label>
    </div>
  </li>
  
</ul>
`;
